x/group/server/msg_server.go

```diff
@@ -0,0 +1,604 @@
+	accountAddress, err := sdk.AccAddressFromBech32(req.GroupAccount)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "request group account")
+	}
+	comment := req.Comment
+	proposers := req.Proposers
+	msgs := req.GetMsgs()
+
+	if err := assertCommentSize(comment, s.maxCommentSize(ctx), "comment"); err != nil {
+		return nil, err
+	}
+
+	account, err := s.getGroupAccountInfo(ctx, accountAddress.Bytes())
+
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "load group account")
+	}
+
+	g, err := s.getGroupInfo(ctx, account.GroupId)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "get group by account")
+	}
+
+	// only members can propose
+	for i := range proposers {
+		if !s.groupMemberTable.Has(ctx, group.GroupMember{GroupId: g.GroupId, Member: proposers[i]}.NaturalKey()) {
+			return nil, sdkerrors.Wrapf(group.ErrUnauthorized, "not in group: %s", proposers[i])
+		}
+	}
+
+	address, err := sdk.AccAddressFromBech32(account.GroupAccount)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group account")
+	}
+
+	if err := ensureMsgAuthZ(msgs, address); err != nil {
+		return nil, err
+	}
+
+	blockTime, err := gogotypes.TimestampProto(ctx.BlockTime())
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "block time conversion")
+	}
+
+	policy := account.GetDecisionPolicy()
+
+	if policy == nil {
+		return nil, sdkerrors.Wrap(group.ErrEmpty, "nil policy")
+	}
+
+	// prevent proposal that can not succeed
+	err = policy.Validate(g)
+	if err != nil {
+		return nil, err
+	}
+
+	timeout := policy.GetTimeout()
+	window, err := gogotypes.DurationFromProto(&timeout)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "maxVotingWindow time conversion")
+	}
+	endTime, err := gogotypes.TimestampProto(ctx.BlockTime().Add(window))
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "end time conversion")
+	}
+
+	m := &group.Proposal{
+		GroupAccount:        req.GroupAccount,
+		Comment:             comment,
+		Proposers:           proposers,
+		SubmittedAt:         *blockTime,
+		GroupVersion:        g.Version,
+		GroupAccountVersion: account.Version,
+		Result:              group.ProposalResultUnfinalized,
+		Status:              group.ProposalStatusSubmitted,
+		ExecutorResult:      group.ProposalExecutorResultNotRun,
+		Timeout:             *endTime,
+		VoteState: group.Tally{
+			YesCount:     "0",
+			NoCount:      "0",
+			AbstainCount: "0",
+			VetoCount:    "0",
+		},
+	}
+	if err := m.SetMsgs(msgs); err != nil {
+		return nil, sdkerrors.Wrap(err, "create proposal")
+	}
+
+	id, err := s.proposalTable.Create(ctx, m)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "create proposal")
+	}
+
+	// TODO: add event?
+
+	return &group.MsgCreateProposalResponse{ProposalId: group.ProposalID(id)}, nil
+}
+
+func (s serverImpl) Vote(ctx types.Context, req *group.MsgVoteRequest) (*group.MsgVoteResponse, error) {
+	id := req.ProposalId
+	voters := req.Voters
+	choice := req.Choice
+	comment := req.Comment
```

Comment:
This is bit confusing. Why `voters` instead of `voter`?

Reply:
It can take as input multiple voters with the same choice per the initial design. I agree this is not ideal and it would be good to be able to define different choice for each voter, or rather to have something similar to `x/gov` where only one voter can vote at a time, wdyt?

Reply:
Yeah, to keep it simple we can just allow one vote for now